### PR TITLE
Throw ContractError in template precondition field

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Part of #8020 

No ghc patch required. :-)

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — COMMENTED** [review]
I’m not quite sure if this is the right approach. It is one thing to rely on desugaring of a handcrafted instance which we create and I don’t really see a way around this.

However, in this case it seems like we’re sidestepping GHC’s instance resolution without a good reason. Generating different code in GHC that calls `show` should just work™ without having to match on the instance here. While that’s more expensive in the short term since changing GHC is annoyin, I think we’re better off with that approach in the long term.

**Comment by sofiafaro-da** [comment]
> I’m not quite sure if this is the right approach. It is one thing to rely on desugaring of a handcrafted instance which we create and I don’t really see a way around this.
> 
> However, in this case it seems like we’re sidestepping GHC’s instance resolution without a good reason. Generating different code in GHC that calls `show` should just work™ without having to match on the instance here. While that’s more expensive in the short term since changing GHC is annoyin, I think we’re better off with that approach in the long term.

I disagree. I don't think we're bypassing instance resolution any more than before. The Show instance is created (via a deriving clause) alongside all these other template instances and we need it for the purpose of implementing the template. It's the same as all these other instances. (We could have hand-crafted the Show instance as well, but that seems unnecessary when `deriving Show` is serviceable.)

We can probably force GHC to perform typeclass resolution in all these cases, and pick this up later during LFConversion. But that would be a whole new mechanism that's unrelated to the existing approach. I don't see the value in changing it just for Show, and I don't think it's worth mixing it with the implementation of exceptions.

**Comment by cocreature** [comment]
> I disagree. I don't think we're bypassing instance resolution any more than before. The Show instance is created (via a deriving clause) alongside all these other template instances and we need it for the purpose of implementing the template. It's the same as all these other instances. (We could have hand-crafted the Show instance as well, but that seems unnecessary when `deriving Show` is serviceable.)

I do think there is a meaningful difference. For the other instances, we need to find them and stuff them into different places or modify them. `ensure` was a complete blackbox to the LF conversion so far and it just stuffed it in the right place. This PR breaks that.
However, I also just realized why my suggestion doesn’t work unfortunately: We need this to be LF version dependent and we cannot have all the logic for that in GHC. I agree that if we have to give up on the whole `ensure` implementation being a blackbox, resolving the `Show` instance doesn’t make things significantly worse :+1:

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 1]
```diff
@@ -694,6 +697,31 @@ convertTemplate env tplTypeCon tbinds@TemplateBinds{..}
...
+            Nothing ->
+                error ("Missing Show instance for template: " <> show tplTypeCon)
+            Just showDict ->
+                ECase b
+                    [ CaseAlternative (CPBool True) ETrue
+                    , CaseAlternative (CPBool False)
+                        $ EThrow TBool (TCon contractErrorTypeCon)
+                        $ mkContractError
+                        $ EBuiltin BEAppendText
+                            `ETmApp` EBuiltin (BEText "Template pre-condition violated: " )
```
[the remark this answers is not in the document]
Comment (sofiafaro-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:713]: I'll change the other use of pre-condition then -- I agree the spelling is off.

**Pull request merged**